### PR TITLE
add getFindAggregation and createIndex methods

### PR DESCRIPTION
**Pull request by andrewtookay**

### CODE REVIEW

File: app/dot/cpp/repository/repository/BaseRepository.java [hunk 1]
```diff
@@ -246,4 +248,11 @@ protected Aggregation<T> getHistoryAggregation(Filter filter) {
                 .getCollection(getEntityType().getSimpleName() + "_history", getEntityType()));
     return filter != null ? historyAggregation.match(filter) : historyAggregation;
   }
+
+  @NotNull
+  public Aggregation<T> getFindAggregation(String id, long timestamp) {
```
Comment (sircoderin) [inline app/dot/cpp/repository/repository/BaseRepository.java]: I find really hard to get "get find". Can't we just use simple getAggregation or getAggregationById?

File: app/dot/cpp/repository/repository/BaseRepository.java [hunk 2]
```diff
@@ -123,8 +125,8 @@ public List<T> listHistoryRecords(String id) {
     final var historyEntities = new ArrayList<T>();
 
     getHistoryCollection()
-        .find(eq(RECORD_ID, id))
-        .sort(descending(TIMESTAMP))
+        .find(Filters.eq(RECORD_ID, id))
```
Comment (sircoderin) [inline app/dot/cpp/repository/repository/BaseRepository.java:128]: let's just use eq all around this class of full name with class. i see in some places you removed the class but here it was added

File: app/dot/cpp/repository/repository/BaseRepository.java [hunk 3]
```diff
@@ -123,8 +125,8 @@ public List<T> listHistoryRecords(String id) {
...
+        .sort(Sorts.descending(TIMESTAMP))
```
Comment (sircoderin) [inline app/dot/cpp/repository/repository/BaseRepository.java:129]: same as above

File: app/dot/cpp/repository/services/RepositoryService.java [hunk 4]
```diff
@@ -32,49 +33,66 @@ public RepositoryService(Config config) {
...
+    for (var entity : entities) {
+      var schema = getSchema(entity, schemaGeneratorConfig);
+      var validationOptions =
+          new ValidationOptions().validator(Filters.jsonSchema(Document.parse(schema)));
 
-          if (withHistory) {
-            createCollection(
-                database, entity.getSimpleName() + "_history", schema, validationOptions);
+      createCollection(database, entity.getSimpleName(), schema, validationOptions);
+      createIndexes(database, index(entity, RECORD_ID, ascending(RECORD_ID), true));
```
Comment (sircoderin) [inline app/dot/cpp/repository/services/RepositoryService.java]: replace with Index.from()

File: app/dot/cpp/repository/services/RepositoryService.java [hunk 5]
```diff
@@ -32,49 +33,66 @@ public RepositoryService(Config config) {
...
 
-            createIndex(database, entity.getSimpleName(), RECORD_ID, true);
-            createIndex(database, entity.getSimpleName() + "_history", RECORD_ID, false);
-          }
+      if (withHistory) {
+        createCollection(database, entity.getSimpleName() + "_history", schema, validationOptions);
+        createIndexes(
+            database,
+            index(entity.getSimpleName() + "_history", RECORD_ID, ascending(RECORD_ID), false));
```
Comment (sircoderin) [inline app/dot/cpp/repository/services/RepositoryService.java]: ^

File: app/dot/cpp/repository/models/Index.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
+package dot.cpp.repository.models;
+
+import org.bson.conversions.Bson;
+
+public class Index {
+
+  public final String collectionName;
+  public final String name;
+  public final Bson spec;
```
Comment (sircoderin) [inline app/dot/cpp/repository/models/Index.java]: what's spec?

**Review by sircoderin — CHANGES_REQUESTED** [review]

File: app/dot/cpp/repository/repository/BaseRepository.java (hunk 2 above)
Comment (andrewtookay) [reply app/dot/cpp/repository/repository/BaseRepository.java:128]: These are Mongo driver Filters and Sorts, not Morphia, that's why they are written differently. I decided to import the Morphia ones statically and these as normal imports.

File: app/dot/cpp/repository/repository/BaseRepository.java (hunk 3 above)
Comment (andrewtookay) [reply app/dot/cpp/repository/repository/BaseRepository.java:129]: ^^

**Pull request merged**